drivers/Makefile.dep

```diff
@@ -36,6 +36,12 @@ ifneq (,$(filter at,$(USEMODULE)))
   USEMODULE += isrpipe_read_timeout
 endif
 
+ifneq (,$(filter at24c%, $(USEMODULE)))
+  FEATURES_REQUIRED += periph_i2c
+  USEMODULE += xtimer
+  USEMODULE += at24cxxx
+endif
+
```

Comment:
Add

``` Makefile
ifneq (,$(filter at24cxxx_eeprom_wrapper, $(USEMODULE)))
  FEATURES_PROVIDED += periph_eeprom
endif
```

here.

Reply:
I think @cladmi won't like this ;)

Reply:
So you suggest to add `drivers/Makefile.features` instead and add below here

https://github.com/RIOT-OS/RIOT/blob/97de3e835b79a303c292a0eaff3eddaa7f06690c/Makefile.dep#L5-L14

This?

``` Makefile
include $(RIOTBASE)/drivers/Makefile.dep
```

Reply:
I would not do it in a `Makefile.features`.
The current difference between `Makefile.dep` and `Makefile.features` is that the `features` only need to be parsed once as they do not depend on the modules.
So if it needs to be done using `USEMODULE` it should be in a `Makefile.dep`.

The current order I am trying to clean and reach is parsing `features` first, then `dep` then `include`. Adding other confusions to this would just be bad.


> ```makefile
> ifneq (,$(filter at24cxxx_eeprom_wrapper,$(USEMODULE)))
>   FEATURES_PROVIDED += periph_eeprom
> endif
> ```


This is something not currently really cleanly handled.
We tend to do the opposite, for the `periph_rtc` that are implemented with `periph_rtt`:
(it is not even consistent)

```
cpu/fe310/Makefile.dep:ifneq (,$(filter periph_rtc,$(USEMODULE)))
cpu/fe310/Makefile.dep-  FEATURES_REQUIRED += periph_rtt
cpu/fe310/Makefile.dep-endif
--
cpu/kinetis/Makefile.dep:ifneq (,$(filter periph_rtc,$(USEMODULE)))
cpu/kinetis/Makefile.dep-  USEMODULE += periph_rtt
cpu/kinetis/Makefile.dep-endif
```
<details><summary>With of course a counter example</summary>

```
boards/esp32-olimex-evb/Makefile.features-# additional features provided by the board (no ADC and no DAC)
boards/esp32-olimex-evb/Makefile.features-ifneq (,$(filter olimex_esp32_gateway,$(USEMODULE)))
boards/esp32-olimex-evb/Makefile.features:    FEATURES_PROVIDED += periph_adc
boards/esp32-olimex-evb/Makefile.features-endif
```
</details>

The boards then define `FEATURES_PROVIDED += periph_rtc` knowing that for this architecture it is implemented through the `rtt`.

The current issue with doing it the other way around, is that now developers request the feature they want and it triggers the modules inclusion.
They do `FEATURES_REQUIRED += periph_eeprom` and the build system finds if it is supported by this board and triggers the inclusion of the correct modules.

And this would currently be problematic as there is currently no way to know what module would provide this feature and do the automatic resolution. (was also discussed in the `netif` refactoring PR https://github.com/RIOT-OS/RIOT/pull/11676)

To implement it the other way around, there would need some handling that triggers using `at24cxxx_eeprom_wrapper`.
For periph, there is some black magic that says all `periph_*` features are implemented by the same module name. The other ones need manual association by checking `FEATURES_USED`

```
cpu/esp32/Makefile.dep:ifneq (,$(filter cpp,$(FEATURES_USED)))
cpu/esp32/Makefile.dep-    USEMODULE += pthread
cpu/esp32/Makefile.dep-    BASELIBS += -lstdc++
cpu/esp32/Makefile.dep-endif
```

Reply:
Since I don't have as much expertise as @cladmi with the build system, some parts of this comment might be wrong.
But after working a bit on #11676, receiving some comments and thinking a bit of features, I think they should be parsed after the dependency tree is known.
My understanding of the goal of provided features is to determine if all required features for a given application are satisfied by the (hardware) setup. If all are not, a big error message in red is printed during the build but this doesn't stop it. So it's just something used to inform the developer that something is missing.
For simplicity, this should allow the dependency tree to be computed before parsing the provided features. And I also think the "features used" has nearly the same semantic as "required features" but it looks like it's just here to justify that provided features are parsed before all dependencies are known (I might be wrong here so would be happy if someone gives a better point of view).

Reply:
Hmm, this makes me question the whole concept of the `FEAUTRES_PROVIDED`, as it would work solely for stuff that **cannot** be provided as a platform independent driver. But for everything that currently is a `periph_*` there are external devices that could be combined with any board externally:

- GPIO: There are I2C GPIO extenders
- SPI: Can be bitbanged (and there even is a driver for that in RIOT)
- I2C: Can be bitbanged
- EEPROM: External I2C modules
- RTC: External modules, e.g. via I2C

And there is **significant** value and demand on using the same API regardless of the interface/device being bundled with an MCU or added later on, as this would allow sharing code using them.

There is work going on to allow using GPIO-Extenders via the GPIO-API: https://github.com/RIOT-OS/RIOT/issues/9690, https://github.com/RIOT-OS/RIOT/issues/9582#issuecomment-410094702. So this issue will arise there again.

Reply:
> But after working a bit on #11676, receiving some comments and thinking a bit of features, I think they should be parsed after the dependency tree is known.
> My understanding of the goal of provided features is to determine if all required features for a given application are satisfied by the (hardware) setup. If all are not, a big error message in red is printed during the build but this doesn't stop it. So it's just something used to inform the developer that something is missing.

It is used to determine compatible applications for `info-boards-supported`. So is used to determine compatible applications.
The fact that it is not an error and lets the build going is for me a stupid side effect from the implementation.

Features are not only hardware requirement.
In the main cases, they are the validation that there is a BSP support (hardware or toolchain), a module to use this hardware, and a pinout mapping of the hardware. (or similar for soft features). My most detailed explanation is currently here https://github.com/RIOT-OS/RIOT/pull/10062#issuecomment-430322739
Using the `at86rf231` module is not enough to have a radio on your board, you need to have the hardware.

> For simplicity, this should allow the dependency tree to be computed before parsing the provided features.
> And I also think the "features used" has nearly the same semantic as "required features" but it looks like it's just here to justify that provided features are parsed before all dependencies are known (I might be wrong here so would be happy if someone gives a better point of view

You are wrong here.

Features define both the "hard" dependency but also allows to currently implement inclusion policy through `FEATURES_OPTIONAL`.

https://github.com/RIOT-OS/RIOT/blob/dd2504852080b2c0a30cc83ce78c2114abe14a1a/Makefile.features#L27-L35

`FEATURES_USED` are not only the required ones, but the result of the policy to include by default all what is optional.

Changing this means changing the policy of resolution. And would require somehow another mechanism for the same thing.
So it will need a dedicated design and explanation on how all is kept or replaced.

Anyway, any of these change is not just the result of a "oh I would like to have this and it would be nice", but an in practice consequence of all what is currently needed, having a working and consistent implementation, propagating the changes to all boards and modules.
The current implementation is not even consistent with what it says it is doing. If you want to help with this there is ongoing work tracked by https://github.com/RIOT-OS/RIOT/issues/9913

I personnaly do not plan to change anything before being consistent, as I would not want to handle the current quirks during a migration.

Reply:
> Hmm, this makes me question the whole concept of the `FEAUTRES_PROVIDED`, as it would work solely for stuff that **cannot** be provided as a platform independent driver. But for everything that currently is a `periph_*` there are external devices that could be combined with any board externally:
> 
>     * GPIO: There are I2C GPIO extenders
> 
>     * SPI: Can be bitbanged (and there even is a driver for that in RIOT)
> 
>     * I2C: Can be bitbanged
> 
>     * EEPROM: External I2C modules
> 
>     * RTC: External modules, e.g. via I2C
> 
> 
> And there is **significant** value and demand on using the same API regardless of the interface/device being bundled with an MCU or added later on, as this would allow sharing code using them.
> 
> There is work going on to allow using GPIO-Extenders via the GPIO-API: #9690, [#9582 (comment)](https://github.com/RIOT-OS/RIOT/issues/9582#issuecomment-410094702). So this issue will arise there again.

It would still work for the "software" implementation, like currently `rtc` over `rtt`.

However, I agree that the current "magic" resolution is really in favor of the hardware ones (I do not like magic).

It would currently require manually handling to implement this behavior, check what is requesting the feature and handle the "FEATURES_OPTIONAL" policy.

Somehow having the possibility to declare "feature A is provided by module M or feature B" is an evolution that makes sense in the build system and would really help with declaring the hardware peripherals (called drivers) with "features", and do the radio thing from https://github.com/RIOT-OS/RIOT/pull/11676
But to have a practical implementation, it may require to have already changed to declarative dependencies instead of the `ifneq` hack.